File: sampleApp/src/main/java/com/zeyad/usecases/app/screens/user/list/UserListVM.java
package com.zeyad.usecases.app.screens.user.list;

import com.zeyad.gadapter.ItemInfo;
import com.zeyad.rxredux.core.redux.BaseEvent;
import com.zeyad.rxredux.core.redux.BaseViewModel;
import com.zeyad.rxredux.core.redux.StateReducer;
import com.zeyad.usecases.api.IDataService;
import com.zeyad.usecases.app.screens.user.list.events.DeleteUsersEvent;
import com.zeyad.usecases.app.screens.user.list.events.GetPaginatedUsersEvent;
import com.zeyad.usecases.app.screens.user.list.events.SearchUsersEvent;
import com.zeyad.usecases.requests.GetRequest;
import com.zeyad.usecases.requests.PostRequest;

import java.util.ArrayList;
import java.util.Collections;
import java.util.HashSet;
import java.util.List;

import io.reactivex.Flowable;
import io.reactivex.Observable;
import io.reactivex.functions.BiFunction;
import io.reactivex.functions.Function;

import static com.zeyad.usecases.app.utils.Constants.URLS.USER;
import static com.zeyad.usecases.app.utils.Constants.URLS.USERS;

/**
 * @author zeyad on 11/1/16.
 */
public class UserListVM extends BaseViewModel<UserListState> {

    private IDataService dataUseCase;

    @Override
    public void init(Object... otherDependencies) {
        if (dataUseCase == null) {
            dataUseCase = (IDataService) otherDependencies[0];
        }
    }

    @Override
    public StateReducer<UserListState> stateReducer() {
        return (newResult, event, currentStateBundle) -> {
            List resultList = (List) newResult;
            List<User> users;
            if (currentStateBundle == null || currentStateBundle.getUsers() == null)
                users = new ArrayList<>();
            else users = Observable.fromIterable(currentStateBundle.getUsers())
                    .map(ItemInfo::<User>getData).toList().blockingGet();
            List<User> searchList = new ArrayList<>();
            switch (event) {
                case "GetPaginatedUsersEvent":
                    users.addAll(resultList);
                    break;
                case "SearchUsersEvent":
                    searchList.addAll(resultList);
                    break;
                case "DeleteUsersEvent":
                    users = Observable.fromIterable(users)
                            .filter(user -> !resultList.contains((long) user.getId()))
                            .distinct()
                            .toList()
                            .blockingGet();
                    break;
                default:
                    break;
            }
            int lastId = users.get(users.size() - 1).getId();
            users = new ArrayList<>(new HashSet<>(users));
            Collections.sort(users, (user1, user2) ->
                    String.valueOf(user1.getId()).compareTo(String.valueOf(user2.getId())));
            return UserListState.builder().users(users).searchList(searchList).lastId(lastId).build();
        };
    }

    @Override
    protected Function<BaseEvent, Flowable<?>> mapEventsToActions() {
        return event -> {
            Flowable action = Flowable.empty();
            if (event instanceof GetPaginatedUsersEvent) {
                action = getUsers(((GetPaginatedUsersEvent) event).getPayLoad());
            } else if (event instanceof DeleteUsersEvent) {
                action = deleteCollection(((DeleteUsersEvent) event).getPayLoad());
            } else if (event instanceof SearchUsersEvent) {
                action = search(((SearchUsersEvent) event).getPayLoad());
            }
            return action;
        };
    }

    public Flowable<User> getUser() {
        return dataUseCase.getObjectOffLineFirst(new GetRequest.Builder(User.class, true)
                .url(String.format(USER, "Zeyad-37"))
                .id("Zeyad-37", User.LOGIN, String.class)
                .cache(User.LOGIN)
                .build());
    }

    public Flowable<List<User>> getUsers(long lastId) {
        return lastId == 0 ?
                dataUseCase.getListOffLineFirst(new GetRequest.Builder(User.class, true)
                        .url(String.format(USERS, lastId))
                        .cache(User.LOGIN)
                        .build()) :
                dataUseCase.getList(new GetRequest.Builder(User.class, true)
                        .url(String.format(USERS, lastId))
                        .build());
    }

    public Flowable<List<User>> search(String query) {
        return dataUseCase.<User>queryDisk(realm -> realm.where(User.class).beginsWith(User.LOGIN, query))
                .zipWith(dataUseCase.<User>getObject(new GetRequest.Builder(User.class, false)
                                .url(String.format(USER, query))
                                .build())
                                .onErrorReturnItem(new User())
                                .filter(user -> user.getId() != 0)
                                .map(user -> user != null ?
                                        Collections.singletonList(user) : Collections.emptyList()),
                        (BiFunction<List<User>, List<User>, List<User>>) (users, singleton) -> {
                            users.addAll(singleton);
                            return new ArrayList<>(new HashSet<>(users));
                        });
    }

    public Flowable<List<String>> deleteCollection(List<String> selectedItemsIds) {
        return dataUseCase.deleteCollectionByIds(new PostRequest.Builder(User.class, true)
                //                .url(USER)
                .payLoad(selectedItemsIds)
                .idColumnName(User.LOGIN, String.class)
                .cache()
                //                .queuable(false, false)
                .build())
                .map(o -> selectedItemsIds);
    }
}

File: sampleApp/src/main/java/com/zeyad/usecases/app/screens/user/list/UserListVM.kt
package com.zeyad.usecases.app.screens.user.list

import com.zeyad.rxredux.core.BaseEvent
import com.zeyad.rxredux.core.viewmodel.BaseViewModel
import com.zeyad.rxredux.core.viewmodel.StateReducer
import com.zeyad.usecases.api.IDataService
import com.zeyad.usecases.app.screens.user.User
import com.zeyad.usecases.app.screens.user.list.events.DeleteUsersEvent
import com.zeyad.usecases.app.screens.user.list.events.GetPaginatedUsersEvent
import com.zeyad.usecases.app.screens.user.list.events.SearchUsersEvent
import com.zeyad.usecases.app.utils.Constants
import com.zeyad.usecases.requests.GetRequest
import com.zeyad.usecases.requests.PostRequest
import io.reactivex.Flowable
import io.reactivex.Observable
import io.reactivex.functions.Function
import java.util.*

/**
 * @author ZIaDo on 6/8/18.
 */
class UserListVM(private var dataUseCase: IDataService) : BaseViewModel<UserListState>() {

    override fun mapEventsToActions(): Function<BaseEvent<*>, Flowable<*>> {
        return Function { event ->
            when (event) {
                is GetPaginatedUsersEvent -> getUsers(event.getPayLoad())
                is DeleteUsersEvent -> deleteCollection(event.getPayLoad())
//                is SearchUsersEvent -> search(event.getPayLoad())
                else -> Flowable.empty<Any>()
            }
        }
    }

    override fun stateReducer(): StateReducer<UserListState> {
        return object : StateReducer<UserListState> {
            override fun reduce(newResult: Any, event: BaseEvent<*>, currentStateBundle: UserListState?): UserListState {
                var users: MutableList<User>
                users = if (currentStateBundle?.users == null)
                    ArrayList<User>()
                else
                    Observable.fromIterable(currentStateBundle.users)
                            .map<User> { it.getData() }
                            .toList().blockingGet()
                val searchList = ArrayList<User>()
                when (event) {
                    is GetPaginatedUsersEvent -> users.addAll(newResult as List<User>)
                    is SearchUsersEvent -> searchList.addAll(newResult as List<User>)
                    is DeleteUsersEvent -> users = Observable.fromIterable(users)
                            .filter { user -> !(newResult as List<*>).contains(user.login) }
                            .distinct().toList().blockingGet()
                    else -> {
                    }
                }
                return UserListState.builder()
                        .users(users)
                        .searchList(searchList)
                        .lastId(users[users.size - 1].id.toLong())
                        .build()
            }
        }
    }

    private fun getUsers(lastId: Long): Flowable<List<User>> {
        val getRequest = GetRequest(String.format(Constants.URLS.USERS, lastId), "",
                User::class.java, Int::class.java, true, "id")
        return if (lastId == 0L)
            dataUseCase.getListOffLineFirst(getRequest)
        else
            dataUseCase.getList(getRequest)
    }

//    private fun search(query: String): Flowable<List<User>> {
//        return dataUseCase
//                .queryDisk(object : RealmQueryProvider<User> {
//                    override fun create(realm: Realm): RealmQuery<User> {
//                        return realm.where(User::class.java).beginsWith(User.LOGIN, query)
//                    }
//                })
//                .map { it.toMutableList() }
//                .zipWith<List<User>>(dataUseCase.getObject<User>(GetRequest.Builder(User::class.java, false)
//                        .url(String.format(USER, query))
//                        .build())
//                        .onErrorReturnItem(User())
//                        .filter { user -> user.id != 0 },
//                        BiFunction<MutableList<User>, User, List<User>> { users, singleton ->
//                            users.add(singleton)
//                            users.toSet().toList()
//                        })
//    }

    private fun deleteCollection(selectedItemsIds: List<String>): Flowable<List<String>> {
        return dataUseCase.deleteCollectionByIds<Boolean>(PostRequest("", "", User::class.java,
                Boolean::class.java, true, false, false, false,
                true, User.LOGIN, String::class.java, `object` = selectedItemsIds))
                .map { selectedItemsIds }
    }
}